certora/specs/Delegation.spec

```diff
@@ -0,0 +1,103 @@
+import "ERC20.spec";
+
+methods {
+    function _moveDelegateVotesExternal(address from, address to, uint256 amount) external envfree;
+    function _burnExternal(address, uint256) external envfree;
+    function ERC20Upgradeable._burn(address account, uint256 value) internal
+        => summaryBurn(account, value);
+    function DelegationToken._moveDelegateVotes(address from, address to, uint256 amount) internal
+        => summaryMove(from, to, amount);
+}
+
+ghost mathint sumOfVotingPower {
+    init_state axiom sumOfVotingPower == 0;
+}
+
+// The ghost variable burnCalled is used to track calls to burn.
+persistent ghost bool burnCalled {
+    init_state axiom burnCalled == false;
+}
+
+function summaryBurn(address a, uint256 amount) {
+    // Mark _burn as being called.
+    burnCalled = true;
+
+    // Burn tokens.
+    _burnExternal(a, amount);
+}
+
+function summaryMove(address from, address to, uint256 amount) {
+    if (burnCalled) {
+        // Ensure that burnt voting power is deducted from sumOfVotingPower.
+        sumOfVotingPower = sumOfVotingPower - amount;
+        burnCalled = false;
+    } else if (from == 0 && to != 0) {
+        // Ensure that voting power is deducted from sumOfVotingPower during a transfer or a delegation when delegatee is zero address.
+        sumOfVotingPower = sumOfVotingPower - amount;
+    } else if (to == 0 && from !=0) {
+        // Ensure that voting power is added to sumOfVotingPower during a transfer or a delegation when delegatee is zero address.
+        sumOfVotingPower = sumOfVotingPower + amount;
+    }
+
+    // Move tokens.
+    _moveDelegateVotesExternal(from, to, amount);
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sload uint256 votingPower (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] {
+    require sumOfVotingPower >= to_mathint(votingPower);
+}
```

Comment:
This is the slot of the `_delegatedVotingPower` mapping. It is never changed by solidity code, instead the targets of this mapping are changed. This may be the reason why you were getting counter-examples

Reply:
Nevermind, I totally missed the "[KEY ...]" :/

Reply:
I know it's hard to read. I tried to use def to make it more readable but it doesn't work for that purpose.